crypto/ec/curve448/curve448utils.h

```diff
@@ -0,0 +1,107 @@
+ * alignment on a given platform to support dynamic linking, since even if you
+ * header was built with eg arch_neon, you might end up linking a library built
+ * with arch_arm32.
+ */
+# ifndef C448_WORD_BITS
+#  if (defined(__ILP64__) || defined(__amd64__) || defined(__x86_64__) \
+      || (((__UINT_FAST32_MAX__)>>30)>>30))
+#   define C448_WORD_BITS 64      /* The number of bits in a word */
+#  else
+#   define C448_WORD_BITS 32      /* The number of bits in a word */
+#  endif
+# endif
+
+# if C448_WORD_BITS == 64
+/* Word size for internal computations */
+typedef uint64_t c448_word_t;
+/* Signed word size for internal computations */
+typedef int64_t c448_sword_t;
+/* "Boolean" type, will be set to all-zero or all-one (i.e. -1u) */
+typedef uint64_t c448_bool_t;
+/* Double-word size for internal computations */
+typedef __uint128_t c448_dword_t;
+/* Signed double-word size for internal computations */
+typedef __int128_t c448_dsword_t;
+# elif C448_WORD_BITS == 32
+/* Word size for internal computations */
+typedef uint32_t c448_word_t;
+/* Signed word size for internal computations */
+typedef int32_t c448_sword_t;
+/* "Boolean" type, will be set to all-zero or all-one (i.e. -1u) */
+typedef uint32_t c448_bool_t;
+/* Double-word size for internal computations */
+typedef uint64_t c448_dword_t;
+/* Signed double-word size for internal computations */
+typedef int64_t c448_dsword_t;
+# else
+#  error "Only supporting C448_WORD_BITS = 32 or 64 for now"
+# endif
+
+/* C448_TRUE = -1 so that C448_TRUE & x = x */
+# define C448_TRUE      (0 - (c448_bool_t)1)
+
+/* C448_FALSE = 0 so that C448_FALSE & x = 0 */
+# define C448_FALSE     0
+
+/* Another boolean type used to indicate success or failure. */
+typedef enum {
+    C448_SUCCESS = -1, /**< The operation succeeded. */
+    C448_FAILURE = 0   /**< The operation failed. */
+} c448_error_t;
+
+/* Return success if x is true */
+static ossl_inline c448_error_t c448_succeed_if(c448_bool_t x)
+{
+    return (c448_error_t) x;
+}
+
+#ifdef OPENSSL_SMALL_FOOTPRINT
+# define FOR_LIMB(_i,_start,_end,_x) \
+    do { \
+        for ((_i) = (_start); (_i) < (_end); (_i)++) \
+            {_x}; \
+    } while (0)
+#else
+/* Use an unrolled loop for performance */
+# define FOR_LIMB(_i, _start, _end, _x) \
+    do { \
+        (_i) = (_start); \
+        do { \
+            switch((_end) - (_i)) { \
+            default: \
+            case 8: {_x}; (_i)++; /* fall through */ \
+            case 7: {_x}; (_i)++; /* fall through */ \
+            case 6: {_x}; (_i)++; /* fall through */ \
+            case 5: {_x}; (_i)++; /* fall through */ \
+            case 4: {_x}; (_i)++; /* fall through */ \
+            case 3: {_x}; (_i)++; /* fall through */ \
+            case 2: {_x}; (_i)++; /* fall through */ \
+            case 1: {_x}; (_i)++; continue; \
+            case 0: break; \
+            } \
+            break; \
+        } while(1); \
+    } while(0)
+#endif
```

Comment:
This implies that compiler is capable of recognizing that this is so to say "constant expression" and not generate things like dispatch table to be used with indirect jump or long chain of conditional branches. But here is a question. If you trust compiler with that, then how come you don't trust compiler to unroll the small[!] loop with constant controlling expression? I mean compilers *are* programmed to make this exact kind of trade-off decision. Besides, it's not actually obvious that unrolled loops are always good for performance. Well, they are likely perform better in benchmark, but in real life by the time you execute piece of code it's more than likely to be evicted from cache and it surely takes longer time to bring in 20KB than 4KB, just to execute it once prior it's evicted till next time around...  BTW, I do observe indirect jumps in code generated with GCC...

Reply:
> What would you expect to be generated here?

I would expect that switch(8) is compiled with no branches whatsoever. But if I can't expect that, and apparently we can't, then trade-off makes no sense. I mean loop would do just as fine. As already implied, there is no reason to feel bad about loops. On high-end processors loop control instructions would cost virtually nothing (modulo occasionally mispredicted branches). Loops can actually be more efficient on most popular platform (a.k.a. Intel). This is because it caches *decoded* [and fused] instructions.

Reply:
When you have a literal in the switch that should of course be easy to optimize the jump away.
Do you have a test case where the code generation is sub-optimal?

Reply:
Let's reconstruct events a little bit. Me: "I'm skeptical about this even with 'constant expressions', and I do note *some* indirect jumps". You: "It's fancy term thing, I would expect indirect jump". Me: "Oh! Looks like we are bound to observe indirect jump, then I do suggest we omit it." To summarize. I didn't actually performed deep analysis, so I don't really have test case. And I'd be reluctant to search for one, because it can be time consuming. Also, I view it of somewhat lesser relevance what particular compiler version does, current or future. [Do note that I'm not saying "*no* relevance", but "*lesser* relevance" :-)] More relevant question [in my mind] is if this (or any particular) piece of code worth the effort of maintaining across *broad* scale of compilers, i.e. without making assumption about specific one or its version. And as already said I'd argue that it's appropriate to rely on compiler to make the trade-off decision about small loop unroll. Because it's commonly recognized optimization technique.

Reply:
Yes, good point.